Replace --aerosols with new apps

WalterKolczynski-NOAA opened the pull request:
**Description**
Removes the `--aerosols` option for setup_expt and replaces it with new apps for aerosols to match the idiom for other components. Currently, ATMA and S2SWA are supported.

**Type of change**
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**How Has This Been Tested?**
- [x] Run experiment and workflow setup on Hera
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

WalterKolczynski-NOAA committed:
Replace --aerosols with new apps

Removes the `--aerosols` option for setup_expt and replaces it with new apps
for aerosols to match the idiom for other components. Currently, ATMA and
S2SWA are supported.

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -257,7 +257,7 @@ def get_workflow(dict_configs, cdump='gdas'):
     tasks = []
 
-    if app in ['S2S', 'S2SW']:
+    if app[0:3] in ['S2S']:
```
```suggestion
    if 'S2S' in app:
```
Any downside to that?

WalterKolczynski-NOAA replied:
I don't think so. Still not used to being able to do things like that with strings.

WalterKolczynski-NOAA replied:
Updated

aerorahul commented on ush/rocoto/setup_workflow_fcstonly.py:
```diff
@@ -357,7 +357,7 @@ def get_workflow(dict_configs, cdump='gdas'):
         dep_dict = {'type': 'task', 'name': f'{cdump}waveinit'}
         deps.append(rocoto.add_dependency(dep_dict))
-        if app not in ['S2S', 'S2SW']:
+        if app[0:3] not in ['S2S']:
```
see above recommendation

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -192,7 +189,19 @@ case "${APP}" in
+    export confignamevarfornems="cpld_wave"
+    source $EXPDIR/config.defaults.s2sw
+    ;;
+  S2SWA)
+    export DO_COUPLED="YES"
+    export DO_WAVE="YES"
+    export DO_OCN="YES"
+    export DO_ICE="YES"
+    export DO_AERO="YES"
+    export CCPP_SUITE="FV3_GFS_v17_coupled_p8"
+    export WAVE_CDUMP="both"
+    export cplwav2atm=".true."
+    export confignamevarfornems="cpld_aero_wave"
```
This is not complete, but something along this lines of compressing to avoid duplication of big blocks.

```suggestion
  S2S |  S2SW | S2SWA)
    export DO_COUPLED="YES"
    if [[ $APP == "S2SW" || ${APP} == "S2SWA" ]]; then
      export DO_WAVE="YES"
      export WAVE_CDUMP="both"
      export cplwav2atm=".true."
      export confignamevarfornems="cpld_wave"
    fi
    if [[ $APP == "S2SWA" ]]; then
      export DO_AERO="YES"
      export confignamevarfornems="cpld_aero_wave"
    fi

```

WalterKolczynski-NOAA replied:
Updated

aerorahul reviewed (commented):
See suggesstions.

WalterKolczynski-NOAA committed:
Update logical tests in setup workflow

Updates logical tests of the app against 'S2S' to use list comprehension
rather than comparing for an exact match against a substring. This is a
more pythonic way of doing the comparison.

WalterKolczynski-NOAA committed:
Update config.base to remove redundant S2S code

A lot of the code for the three S2S apps was duplicate, so compressed all
into a single case match, which then sets a few additional things by checking
whether the app also specifies waves and/or aerosols.

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -154,46 +154,47 @@ export DO_COUPLED="NO"
-    export confignamevarfornems="cpld${aero_nems_string}_wave"
+    export confignamevarfornems="cpld"
+
+    if [[ $APP =~ 'A$' ]]; then
+        export DO_AERO="YES"
+        export confignamevarfornems="${confignamevarfornems}_aero"
+    fi
```
This does not work.

A test script:
```
#!/bin/bash

DO_AERO='unknown'
APP=S2SWA
if [[ $APP =~ 'A$' ]]; then
  DO_AERO="YES"
fi

echo "DO_AERO is $DO_AERO"
```
gives:
```
DO_AERO is unknown
```

WalterKolczynski-NOAA replied:
This was a mistake on my part. I had tested this myself and found I had to remove the quotes from the regex, but I never made those changes to the original in the repo. Should work now.

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -154,46 +154,47 @@ export DO_COUPLED="NO"
+        export confignamevarfornems="${confignamevarfornems}_aero"
+    fi
+
+    if [[ $APP =~ '^S2SW' ]]; then
+        export DO_WAVE="YES"
+        export WAVE_CDUMP="both"    
+        export cplwav2atm=".true."
+        export confignamevarfornems="${confignamevarfornems}_wave"
+    fi
```
Neither does this.

```
#!/bin/bash

DO_WAVE='unknown'
if [[ $APP =~ '^S2SW' ]]; then
  DO_WAVE="YES"
fi

echo "DO_WAVE is $DO_WAVE"
```
gives
```
DO_WAVE is unknown
```

WalterKolczynski-NOAA replied:
This one should work now too.

aerorahul reviewed (changes requested):
The logic for the `APP` in `config.base` is broken.

WalterKolczynski-NOAA committed:
Fix string regex of APP in config.base

The string regex operator doesn't work properly if the regex is enclosed
in quotes. Those are now removed.

aerorahul reviewed (approved):
Looks good.

aerorahul commented:
Please merge and send an announcement when ready.
